File: java/amp-eventing/src/main/java/amp/eventing/DigitalSignatureProcessor.java
package amp.eventing;


import java.security.Signature;
import java.security.cert.X509Certificate;

import amp.bus.security.CredentialHolder;
import amp.bus.security.ICertificateProvider;
import amp.bus.security.IUserInfoRepository;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import amp.messaging.EnvelopeHelper;
import amp.messaging.IContinuationCallback;
import amp.messaging.IMessageProcessor;
import amp.messaging.MessageContext;
import amp.messaging.MessageContext.Directions;
import amp.messaging.MessageException;


public class DigitalSignatureProcessor implements IMessageProcessor {

	protected ICertificateProvider certProvider;
	protected IUserInfoRepository userInfoRepo;
	protected CredentialHolder credentials;
	protected Logger log;
	
	
	public DigitalSignatureProcessor(ICertificateProvider certProvider, IUserInfoRepository userInfoRepository) throws Exception {
		this.certProvider = certProvider;
		this.userInfoRepo = userInfoRepository;
		
		this.log = LoggerFactory.getLogger(this.getClass());
		
		try {
			this.credentials = certProvider.getCredentials();
		}
		catch(Exception ex)
		{
			log.error("Failed to get client's PKI credentials", ex);
			throw ex;
		}
	}
	
	
	@Override
	public void processMessage(MessageContext context, IContinuationCallback continuation) throws MessageException {
		
		if (Directions.In == context.getDirection()) {
			this.processInbound(context, continuation);
		}
		else if (Directions.Out == context.getDirection()) {
			this.processOutbound(context, continuation);
		}
	}
	
	public void processOutbound(MessageContext context, IContinuationCallback continuation) throws MessageException {
		
		try {
			EnvelopeHelper env = new EnvelopeHelper(context.getEnvelope());
			
			Signature instance = Signature.getInstance("SHA1withRSA");
			instance.initSign(this.credentials.getPrivateKey());
			instance.update(env.getPayload());			
			
			env.setDigitalSignature(instance.sign());
			
			// make sure that the sender identity has spaces between LDAP elements
			String sender = this.credentials.getCertificate().getSubjectX500Principal().getName();
			sender = sender.replaceAll(",", ", ");
			
			env.setSenderIdentity(sender);
			continuation.continueProcessing();
		}
		catch(Exception ex) {
			log.error("Exception while signing outbound event", ex);
			throw new MessageException("Exception while signing outbound event", ex);
		}
	}

	public void processInbound(MessageContext context, IContinuationCallback continuation) throws MessageException  {

		boolean verified = false;
		
		try {
			EnvelopeHelper env = new EnvelopeHelper(context.getEnvelope());
			String senderIdentity = env.getSenderIdentity();
			
			// remove spaces from the sender identity
			senderIdentity = senderIdentity.replace(", ", ",");
			
			X509Certificate senderCert = this.userInfoRepo.getPublicCertificateFor(senderIdentity);
			
			Signature instance = Signature.getInstance("SHA1withRSA");
			instance.initVerify(senderCert);
			instance.update(env.getPayload());
			
			verified = instance.verify(env.getDigitalSignature());
			
			if (false == verified) {
				log.warn("Event may have been tampered with (id:{})", env.getMessageId().toString());
			}
		}
		catch(Exception ex) {
			log.error("Failed to verify sender's digital signature", ex);
			throw new MessageException("Failed to verify sender's digital signature", ex);
		}
		
		if (verified) { continuation.continueProcessing(); }
	}
}

File: dotnet/source/amp.eventing/DigitalSignatureProcessor.cs
using System;
using System.Security.Cryptography;
using System.Security.Cryptography.X509Certificates;
using amp.messaging;
using Common.Logging;

using cmf.bus;
using amp.bus.security;

namespace amp.eventing
{
    public class DigitalSignatureProcessor : IMessageProcessor
    {
        protected ICertificateProvider _certProvider;
        protected X509Certificate2 _cert;
        protected ILog _log;


        public DigitalSignatureProcessor(ICertificateProvider certProvider)
        {
            _certProvider = certProvider;

            try
            {
                _cert = certProvider.GetCertificate();
            }
            catch (Exception ex)
            {
                _log.Error("Failed to get a certificate from the provider", ex);
            }
        }


        public void ProcessMessage(MessageContext context, Action continueProcessing)
        {
            if (MessageContext.Directions.In == context.Direction)
            {
                this.ProcessInbound(context, continueProcessing);
            }
            if (MessageContext.Directions.Out == context.Direction)
            {
                this.ProcessOutbound(context, continueProcessing);
            }
        }

        public void ProcessInbound(MessageContext context, Action continueProcessing)
        {
            bool signatureVerified = false;
            Envelope env = context.Envelope;

            // start by getting the purported sender's identity
            string sender = env.GetSenderIdentity();
            if (string.IsNullOrEmpty(sender))
            {
                _log.Error("An inbound event arrived with no sender identity");
            }
            else
            {
                // use the identity to lookup their certificate
                X509Certificate2 senderCert = _certProvider.GetCertificateFor(sender);
                if (null == senderCert)
                {
                    _log.Error("Sender " + sender + " does not have a public key certificate available");
                }
                else
                {
                    // get the digital signature from the headers
                    byte[] digitalSignature = env.GetDigitalSignature();
                    if (0 == digitalSignature.LongLength)
                    {
                        _log.Error("Sender " + sender + " did not digitally sign the event");
                    }
                    else
                    {
                        // verify that the payload hasn't been tampered with by using the 
                        // sender's public key and the digital signature on the envelope
                        RSACryptoServiceProvider rsaProvider = senderCert.PublicKey.Key as RSACryptoServiceProvider;
                        signatureVerified = rsaProvider.VerifyData(env.Payload, new SHA1CryptoServiceProvider(), digitalSignature);
                    }
                }
            }

            if (signatureVerified) { continueProcessing(); }
        }

        public void ProcessOutbound(MessageContext context, Action continueProcessing)
        {
            if (null == _cert) { return; }

            Envelope env = context.Envelope;
            env.SetSenderIdentity(_cert.Subject);

            try
            {
                RSACryptoServiceProvider rsaProvider = _cert.PrivateKey as RSACryptoServiceProvider;
                env.SetDigitalSignature(rsaProvider.SignData(env.Payload, new SHA1CryptoServiceProvider()));

                continueProcessing();
            }
            catch (Exception ex)
            {
                _log.Error("Failed to digitally sign the event", ex);
                throw;
            }
        }

        public void Dispose()
        {
            //Nothing to do.
        }
    }
}
